pkg/appgateserver/rings.go

```diff
@@ -0,0 +1,135 @@
+// TODO(@h5law): Move all this logic out into a shared package
+package appgateserver
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	ring_secp256k1 "github.com/athanorlabs/go-dleq/secp256k1"
+	ringtypes "github.com/athanorlabs/go-dleq/types"
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
+	accounttypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	ring "github.com/noot/ring-go"
+
+	"github.com/pokt-network/poktroll/pkg/signer"
+	apptypes "github.com/pokt-network/poktroll/x/application/types"
+)
+
+// getRingSingerForAppAddress returns the RingSinger used to sign relays.
+// This method first attempts to get the points of the ring from the cache, if it
+// fails it queries the application module for the points and creates the ring.
+func (app *appGateServer) getRingSingerForAppAddress(ctx context.Context, appAddress string) (*signer.RingSigner, error) {
+	var ring *ring.Ring
+	var err error
+
+	// lock the cache for reading
+	app.ringCacheMutex.RLock()
+	defer app.ringCacheMutex.RUnlock()
+
+	// check if the ring is in the cache
+	points, ok := app.ringCache[appAddress]
+	if !ok {
+		// if the ring is not in the cache, get it from the application module
+		log.Printf("DEBUG: No ring cached for address: %s", appAddress)
+		ring, err = app.getRingForAppAddress(ctx, appAddress)
+	} else {
+		// if the ring is in the cache, create it from the points
+		log.Printf("DEBUG: Ring cached for address: %s", appAddress)
+		ring, err = newRingFromPoints(points)
+	}
+	if err != nil {
+		log.Printf("ERROR: Unable to get ring for address: %s [%w]", appAddress, err)
+		return nil, err
+	}
+
+	// return the ring signer
+	return signer.NewRingSigner(ring, app.signingInformation.SigningKey), nil
+}
+
+// getRingForAppAddress returns the RingSinger used to sign relays. It does so by fetching
+// the latest information from the application module and creating the correct ring.
+// This method also caches the ring's public keys for future use.
+func (app *appGateServer) getRingForAppAddress(ctx context.Context, appAddress string) (*ring.Ring, error) {
+	points, err := app.getDelegatedPubKeysForAddress(ctx, appAddress)
+	if err != nil {
+		return nil, err
+	}
+	return newRingFromPoints(points)
+}
+
+// newRingFromPoints creates a new ring from a slice of points on the secp256k1 curve
+func newRingFromPoints(points []ringtypes.Point) (*ring.Ring, error) {
+	return ring.NewFixedKeyRingFromPublicKeys(ring_secp256k1.NewCurve(), points)
+}
+
+// getDelegatedPubKeysForAddress returns the ring used to sign a message for the given application
```

Comment:
The name / comment of this function is a bit confusing because we're not doing any delegation. It's just the app and itself.

@h5law  Can you expand or update appropriately?

Reply:
We are getting the delegated pub keys for the application. We add the app twice incase the app is not delegated to anyone but ultimately this will build the ring with all the delegated gateways addresses too